File: packages/react-native-nitro-filetoolkit/android/src/main/java/com/margelo/nitro/filetoolkit/HybridFileWriter.kt
package com.margelo.nitro.filetoolkit

import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.RandomAccessFile

internal class HybridFileWriter(
  override val destination: FileLocation,
  private val destinationFile: File,
  private val mode: WriteMode,
  private val atomicity: Atomicity,
  private val metadata: FileMetadataMapper,
) : HybridFileWriterSpec() {
  private val lock = Any()
  private val stagingFile = FileOperations.siblingStagingFile(destinationFile)
  private val handle: RandomAccessFile
  private var currentPosition = 0L
  private var closed = false
  private var finished = false

  init {
    if (mode == WriteMode.CREATE_NEW && metadata.exists(destinationFile)) {
      throw FileToolkitException.invalidOperation("destination already exists")
    }
    if (mode == WriteMode.APPEND && metadata.exists(destinationFile)) {
      FileInputStream(destinationFile).use { input ->
        FileOutputStream(stagingFile).use { output -> input.copyTo(output) }
      }
    } else if (!stagingFile.createNewFile()) {
      throw FileToolkitException.invalidOperation("cannot create staging file")
    }
    handle = RandomAccessFile(stagingFile, "rw")
    currentPosition = if (mode == WriteMode.APPEND) handle.length() else 0L
    handle.seek(currentPosition)
  }

  override val position: ULong
    get() = synchronized(lock) { currentPosition.toULong() }

  override fun write(data: ArrayBuffer): Promise<Unit> {
    val ownedBytes = data.toByteArray()
    return Promise.parallel {
      synchronized(lock) {
        requireWritable()
        handle.write(ownedBytes)
        currentPosition += ownedBytes.size
      }
    }
  }

  override fun flush(): Promise<Unit> = Promise.parallel {
    synchronized(lock) {
      requireWritable()
      handle.fd.sync()
    }
  }

  override fun commit(): Promise<FileInfo> = Promise.parallel {
    synchronized(lock) {
      requireWritable()
      handle.fd.sync()
      handle.close()
      closed = true
      if (mode == WriteMode.CREATE_NEW) {
        FileOperations.installNew(stagingFile, destinationFile)
      } else {
        FileOperations.atomicReplace(stagingFile, destinationFile, atomicity)
      }
      finished = true
      metadata.info(destinationFile)
    }
  }

  override fun abort(): Promise<Unit> = Promise.parallel {
    synchronized(lock) { finishWithoutCommit() }
  }

  override fun close() {
    synchronized(lock) { finishWithoutCommit() }
  }

  private fun finishWithoutCommit() {
    if (!closed) {
      handle.close()
      closed = true
    }
    if (!finished && stagingFile.exists()) stagingFile.delete()
    finished = true
  }

  private fun requireWritable() {
    if (closed || finished) throw FileToolkitException.invalidOperation("writer is closed")
  }
}

File: packages/react-native-nitro-filetoolkit/ios/HybridFileWriter.swift
import Foundation
import NitroModules

final class HybridFileWriter: HybridFileWriterSpec {
  let destination: FileLocation

  private let queue = DispatchQueue(label: "com.margelo.nitro.filetoolkit.writer", qos: .utility)
  private let lock = NSLock()
  private let destinationURL: URL
  private let stagingURL: URL
  private let mode: WriteMode
  private let atomicity: Atomicity
  private let fileManager: FileManager
  private let metadata: FileMetadataMapper
  private let handle: FileHandle
  private var currentPosition: UInt64
  private var isClosed = false
  private var isFinished = false

  var position: UInt64 {
    lock.lock()
    defer { lock.unlock() }
    return currentPosition
  }

  init(
    destination: FileLocation,
    destinationURL: URL,
    mode: WriteMode,
    atomicity: Atomicity,
    fileManager: FileManager,
    metadata: FileMetadataMapper
  ) throws {
    self.destination = destination
    self.destinationURL = destinationURL
    self.mode = mode
    self.atomicity = atomicity
    self.fileManager = fileManager
    self.metadata = metadata
    stagingURL = FileOperations.siblingStagingURL(for: destinationURL)

    if mode == .createNew && fileManager.fileExists(atPath: destinationURL.path) {
      throw FileToolkitError.invalidOperation("destination already exists")
    }
    if mode == .append && fileManager.fileExists(atPath: destinationURL.path) {
      try fileManager.copyItem(at: destinationURL, to: stagingURL)
    } else {
      guard fileManager.createFile(atPath: stagingURL.path, contents: nil) else {
        throw FileToolkitError.invalidOperation("cannot create staging file")
      }
    }

    handle = try FileHandle(forWritingTo: stagingURL)
    if mode == .append {
      currentPosition = try handle.seekToEnd()
    } else {
      currentPosition = 0
    }
    super.init()
  }

  func write(data: ArrayBuffer) throws -> Promise<Void> {
    let ownedData = data.toData(copyIfNeeded: true)
    return Promise.parallel(queue) { [self] in
      lock.lock()
      defer { lock.unlock() }
      try requireWritable()
      try handle.write(contentsOf: ownedData)
      currentPosition += UInt64(ownedData.count)
    }
  }

  func flush() throws -> Promise<Void> {
    Promise.parallel(queue) { [self] in
      lock.lock()
      defer { lock.unlock() }
      try requireWritable()
      try handle.synchronize()
    }
  }

  func commit() throws -> Promise<FileInfo> {
    Promise.parallel(queue) { [self] in
      lock.lock()
      defer { lock.unlock() }
      try requireWritable()
      try handle.synchronize()
      try handle.close()
      isClosed = true

      if mode == .createNew && fileManager.fileExists(atPath: destinationURL.path) {
        try? fileManager.removeItem(at: stagingURL)
        isFinished = true
        throw FileToolkitError.invalidOperation("destination already exists")
      }
      if mode == .createNew {
        try FileOperations.installNewStagingItem(
          stagingURL,
          destination: destinationURL,
          fileManager: fileManager
        )
      } else {
        try FileOperations.replaceStagingItem(
          stagingURL,
          destination: destinationURL,
          atomicity: atomicity,
          fileManager: fileManager
        )
      }
      isFinished = true
      return try metadata.info(for: destinationURL)
    }
  }

  func abort() throws -> Promise<Void> {
    Promise.parallel(queue) { [self] in
      lock.lock()
      defer { lock.unlock() }
      if !isClosed {
        try handle.close()
        isClosed = true
      }
      if !isFinished && fileManager.fileExists(atPath: stagingURL.path) {
        try fileManager.removeItem(at: stagingURL)
      }
      isFinished = true
    }
  }

  func close() throws {
    lock.lock()
    defer { lock.unlock() }
    guard !isClosed else { return }
    try handle.close()
    isClosed = true
    if !isFinished && fileManager.fileExists(atPath: stagingURL.path) {
      try? fileManager.removeItem(at: stagingURL)
    }
    isFinished = true
  }

  private func requireWritable() throws {
    if isClosed || isFinished {
      throw FileToolkitError.invalidOperation("writer is closed")
    }
  }

  deinit {
    try? handle.close()
    if !isFinished {
      try? fileManager.removeItem(at: stagingURL)
    }
  }
}
